### PR TITLE
MG-2359 - Replace Vault with Absmach certs

### LINKED ISSUE #2359: Add new certificate manager in Certs service
### Is your feature request related to a problem? Please describe.

Vault have changed their license.

### Describe the feature you are requesting, as well as the possible use case(s) for it.

it is good have support multiple certificate managers or we should write our own according to our need 
It should features like 
- Root CA , Intermediate CA, certificate management like generation, rotation, revocation  , etc ...
- OSCP 
- CRL 

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nyagamunene**
# What type of PR is this?
This is a feature because it adds the following functionality: it replaces vault with absmach certs as the PKI manager.

## What does this do?
It replaces vault with absmach certs.

## Which issue(s) does this PR fix/relate to?

- Related Issue #2359
- Resolves #2359

## Have you included tests for your changes?
Yes

## Did you document any new/modified feature?
No

### CODE REVIEW

File: pkg/sdk/go/certs.go [hunk 1]
```diff
@@ -20,33 +20,38 @@ const (
...
-	CertSerial string    `json:"cert_serial,omitempty"`
-	ClientKey  string    `json:"client_key,omitempty"`
-	ClientCert string    `json:"client_cert,omitempty"`
-	Expiration time.Time `json:"expiration,omitempty"`
+	SerialNumber string    `json:"serial_number,omitempty"`
+	Certificate  string    `json:"certificate,omitempty"`
+	Key          string    `json:"key,omitempty"`
+	Revoked      bool      `json:"revoked,omitempty"`
+	ExpiryDate   time.Time `json:"expiry_date,omitempty"`
+	EntityID     string    `json:"entity_id,omitempty"`
```
Comment (dborovcanin) [inline pkg/sdk/go/certs.go]: We can keep ThingID here, just take care you serialize it properly (JSON tag should be `entity_id` and give a comment explaining the mapping.

File: pkg/sdk/go/certs.go [hunk 2]
```diff
@@ -20,33 +20,38 @@ const (
...
-	ThingID    string    `json:"thing_id,omitempty"`
-	CertSerial string    `json:"cert_serial,omitempty"`
-	ClientKey  string    `json:"client_key,omitempty"`
-	ClientCert string    `json:"client_cert,omitempty"`
-	Expiration time.Time `json:"expiration,omitempty"`
+	SerialNumber string    `json:"serial_number,omitempty"`
+	Certificate  string    `json:"certificate,omitempty"`
+	Key          string    `json:"key,omitempty"`
+	Revoked      bool      `json:"revoked,omitempty"`
+	ExpiryDate   time.Time `json:"expiry_date,omitempty"`
```
Comment (dborovcanin) [inline pkg/sdk/go/certs.go]: This is not a date, please change it to `Expiry` or `ExpiryTime`. The same goes for JSON tag name.

File: pkg/sdk/go/certs.go (hunk 1 above)
Comment (dborovcanin) [reply pkg/sdk/go/certs.go]: Update this in the rest of the codebase and API.

File: pkg/sdk/go/certs.go (hunk 2 above)
Comment (dborovcanin) [reply pkg/sdk/go/certs.go]: Update this in the rest of the codebase and API.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — CHANGES_REQUESTED** [review]
@nyagamunene Please rebase. @arvindh123 Please review.

File: docker/addons/certs/docker-compose.yml [hunk 3]
```diff
@@ -88,3 +78,45 @@ services:
...
+      AM_CERTS_DB_SSL_MODE: ${MG_CERTS_DB_SSL_MODE}
+      AM_CERTS_HTTP_HOST: magistrala-am-certs
+      AM_CERTS_HTTP_PORT: 9010
+      AM_CERTS_GRPC_HOST: magistrala-am-certs
+      AM_CERTS_GRPC_PORT: 7012
+      AM_JAEGER_URL: ${MG_JAEGER_URL}
+      AM_JAEGER_TRACE_RATIO: ${MG_JAEGER_TRACE_RATIO}
+    ports:
+      - 9010:9010
+      - 7012:7012
```
Comment (felixgateru) [inline docker/addons/certs/docker-compose.yml]: Add blank line

File: tools/config/mockery.yaml [hunk 4]
```diff
@@ -18,3 +18,10 @@ packages:
           dir: "./auth/mocks"
           mockname: "PolicyServiceClient"
           filename: "policy_client.go"
+  github.com/absmach/magistrala/certs/pki/am-certs:
+    interfaces:
+      Agent:
+        config:
+          dir: "./certs/mocks"
+          mockname: "Agent"
+          filename: "pki.go"
```
Comment (felixgateru) [inline tools/config/mockery.yaml]: Add blank line

File: certs/service.go [hunk 5]
```diff
@@ -127,80 +100,76 @@ func (cs *certsService) RevokeCert(ctx context.Context, token, thingID string) (
...
+		Certificates: certs,
+	}, nil
+}
 
-	cert, err := cs.certsRepo.RetrieveBySerial(ctx, u.GetId(), serialID)
+func (cs *certsService) ViewCert(ctx context.Context, token, serialID string) (sdk.Certificate, error) {
+	_, err := cs.auth.Identify(ctx, &magistrala.IdentityReq{Token: token})
 	if err != nil {
-		return Cert{}, errors.Wrap(svcerr.ErrViewEntity, err)
+		return sdk.Certificate{}, errors.Wrap(svcerr.ErrAuthentication, err)
```
Comment (felixgateru) [inline certs/service.go]: Identify returns an svcerr.ErrAuthentication, so this is repeated.  You can revert

**Review by felixgateru — CHANGES_REQUESTED** [review]

File: certs/service_test.go [hunk 6]
```diff
@@ -37,274 +34,231 @@ const (
...
 			thingID:     "2",
-			page:        certs.Page{},
-			identifyRes: &magistrala.IdentityRes{Id: validID},
+			page:        amsdk.CertificatePage{},
+			identifyRes: &magistrala.IdentityRes{},
 			thingErr:    errors.NewSDKError(certs.ErrFailedCertCreation),
 			err:         certs.ErrFailedCertRevocation,
 		},
+		{
+			desc:        "revoke cert failed list certs",
```
Comment (felixgateru) [inline certs/service_test.go]:
```suggestion
			desc:        "revoke cert with failed to list certs",
```

File: certs/service_test.go [hunk 7]
```diff
@@ -356,93 +310,75 @@ func TestListSerials(t *testing.T) {
...
+		// },
+		// {
+		// 	desc:        "list cert with invalid serial",
+		// 	token:       token,
+		// 	serialID:    invalid,
+		// 	cert:        certs.Cert{},
+		// 	identifyRes: &magistrala.IdentityRes{Id: validID},
+		// 	repoErr:     repoerr.ErrNotFound,
+		// 	err:         svcerr.ErrNotFound,
+		// },
```
Comment (felixgateru) [inline certs/service_test.go]: uncomment the code

File: tools/config/.golangci.yml [hunk 8]
```diff
@@ -66,7 +66,6 @@ linters:
     - gocritic
     - gosimple
     - errcheck
-    - govet
```
Comment (felixgateru) [inline tools/config/.golangci.yml]: why did you remove this?

File: tools/config/mockery.yaml [hunk 9]
```diff
@@ -18,3 +18,10 @@ packages:
           dir: "./auth/mocks"
           mockname: "PolicyServiceClient"
           filename: "policy_client.go"
+  github.com/absmach/magistrala/certs/pki/am-certs:
```
Comment (felixgateru) [inline tools/config/mockery.yaml]: why did you put this here?

Comment (nyagamunene) [reply tools/config/mockery.yaml]: Its a much better way to generate mocks compared to `go generate`. I have used this because I could easily select the folder where the mock was being saved.

**Review by felixgateru — COMMENTED** [review]
is it possible to view revoked certs

**Comment by nyagamunene** [comment]
> is it possible to view revoked certs

You should be able to view with the updated code.

File: certs/api/transport.go [hunk 10]
```diff
@@ -24,6 +24,8 @@ const (
 	contentType = "application/json"
 	offsetKey   = "offset"
 	limitKey    = "limit"
+	revokeKey   = "revoke"
```
Comment (felixgateru) [inline certs/api/transport.go]: could we try revoked

**Review by felixgateru — CHANGES_REQUESTED** [review]

**Review by felixgateru — DISMISSED** [review]
LGTM

File: certs/api/responses.go [hunk 11]
```diff
@@ -20,12 +20,28 @@ type certsPageRes struct {
 }
 
 type certsRes struct {
-	ThingID    string    `json:"thing_id"`
-	ClientCert string    `json:"client_cert"`
-	ClientKey  string    `json:"client_key"`
-	CertSerial string    `json:"cert_serial"`
-	Expiration time.Time `json:"expiration"`
-	created    bool
+	EntityID     string    `json:"entity_id"`
```
Comment (dborovcanin) [inline certs/api/responses.go]: We are talking about Things here, so let's call it `thing_id`.

File: certs/api/endpoint.go [hunk 12]
```diff
@@ -20,16 +20,11 @@ func issueCert(svc certs.Service) endpoint.Endpoint {
...
 
-		return certsRes{
-			CertSerial: res.Serial,
-			ThingID:    res.ThingID,
-			ClientCert: res.ClientCert,
-			ClientKey:  res.ClientKey,
-			Expiration: res.Expire,
-			created:    true,
+		return serialRes{
+			Serial: res.SerialNumber,
```
Comment (dborovcanin) [inline certs/api/endpoint.go]: We should return all the certificate data except the key. The same goes for Certs service endpoint.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by arvindh123 — CHANGES_REQUESTED** [review]
If certs is not going to have repo, Then we can remove the certs service, and convert it to certs library

File: certs/pki/am-certs/am_certs.go [hunk 13]
```diff
@@ -0,0 +1,136 @@
+// Copyright (c) Abstract Machines
+// SPDX-License-Identifier: Apache-2.0
+package sdk
```
Comment (arvindh123) [inline certs/pki/am-certs/am_certs.go]: This should be am_certs 

Please rename this package name without `_`

File: certs/pki/am-certs/am_certs.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+import (
+	"github.com/absmach/certs/sdk"
+	"golang.org/x/crypto/ocsp"
+)
+
+type Agent interface {
+	Issue(entityId, ttl string, ipAddrs []string) (sdk.Certificate, error)
```
Comment (arvindh123) [inline certs/pki/am-certs/am_certs.go]: The return certificate struct should be belongs to magistrala certs service. 
So please convert it magistrala certs object.

File: certs/service.go [hunk 15]
```diff
@@ -34,35 +35,33 @@ var _ Service = (*certsService)(nil)
 //go:generate mockery --name Service --output=./mocks --filename service.go --quiet --note "Copyright (c) Abstract Machines"
 type Service interface {
 	// IssueCert issues certificate for given thing id if access is granted with token
-	IssueCert(ctx context.Context, token, thingID, ttl string) (Cert, error)
+	IssueCert(ctx context.Context, token, thingID, ttl string) (sdk.Certificate, error)
```
Comment (arvindh123) [inline certs/service.go]: Certs should have it own Type, 

So please convert it magistrala certs object.   
I mean please revert back to original 
In `amcerts` convert sdk.Certificate to magistrala.Cert

File: certs/pki/vault/vault.go [hunk 16]
```diff
@@ -45,8 +45,6 @@ type Cert struct {
 }
 
 // Agent represents the Vault PKI interface.
-//
-//go:generate mockery --name Agent --output=../mocks --filename pki.go --quiet --note "Copyright (c) Abstract Machines"
 type Agent interface {
 	// IssueCert issues certificate on PKI
```
Comment (arvindh123) [inline certs/pki/vault/vault.go:49]: Agent interface should be defined in common at `certs/agent.go`

And both Abstract Machines certs and Vault should use same Agent interface

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: certs/pki/vault/vault.go (hunk 16 above)
Comment (nyagamunene) [reply certs/pki/vault/vault.go:49]: This is going to be a challenge because the methods in the interface are very different.

Comment (arvindh123) [reply certs/pki/vault/vault.go:49]: In present Magistrala , cert service requires only following `IssueCert`, `Read`, `Revoke`,  from certificate provide 
`LoginAndRenew` be an empty function

Comment (arvindh123) [reply certs/pki/vault/vault.go:49]: Why we have remove the certs repo? Just trying to understand  the reason

File: cmd/certs/main.go [hunk 17]
```diff
@@ -63,6 +59,11 @@ type config struct {
 	PkiNamespace string `env:"MG_CERTS_VAULT_NAMESPACE"          envDefault:""`
 	PkiPath      string `env:"MG_CERTS_VAULT_THINGS_CERTS_PKI_PATH"       envDefault:"pki_int"`
 	PkiRole      string `env:"MG_CERTS_VAULT_THINGS_CERTS_PKI_ROLE_NAME"  envDefault:"magistrala"`
```
Comment (felixgateru) [inline cmd/certs/main.go:61]: Do you still need these?

File: certs/service.go [hunk 18]
```diff
@@ -127,80 +107,125 @@ func (cs *certsService) RevokeCert(ctx context.Context, token, thingID string) (
...
+	case "all":
+		for _, c := range cp.Certificates {
+			certs = append(certs, Cert{
+				SerialNumber: c.SerialNumber,
+				ThingID:      c.ThingID,
+				ExpiryTime:   c.ExpiryTime,
+				Revoked:      c.Revoked,
+			})
+		}
 	}
```
Comment (felixgateru) [inline certs/service.go:202]: Would  you consider refactoring to something like this:

```
var certs []Cert
for _, c := range cp.Certificates {
		if (pm.Revoked == "true" && c.Revoked) || (pm.Revoked == "false" && !c.Revoked) || (pm.Revoked == "all") {
			certs = append(certs, Cert{
				SerialNumber: c.SerialNumber,
				ThingID:      c.ThingID,
				ExpiryTime:   c.ExpiryTime,
				Revoked:      c.Revoked,
			})
		}
	}
```

File: cmd/certs/main.go (hunk 17 above)
Comment (felixgateru) [inline cmd/certs/main.go]: Are these still required?

File: certs/service.go (hunk 18 above)
Comment (felixgateru) [inline certs/service.go]: Would you consider something like this:

```
var certs []Cert
for _, c := range cp.Certificates {
		if (pm.Revoked == "true" && c.Revoked) || (pm.Revoked == "false" && !c.Revoked) || (pm.Revoked == "all") {
			certs = append(certs, Cert{
				SerialNumber: c.SerialNumber,
				ThingID:      c.ThingID,
				ExpiryTime:   c.ExpiryTime,
				Revoked:      c.Revoked,
			})
		}
	}
```

**Review by felixgateru — COMMENTED** [review]
The response from issue cert is missing a thingID:
`
{
    "thing_id": "",
    "serial_number": "212000476178773257560355903916569281088",
    "expiry_time": "0001-01-01T00:00:00Z",
    "revoked": false
}
`

**Review by dborovcanin — COMMENTED** [review]
Please check the CI code coverage report and update tests where it makes sense.

**Review by dborovcanin — APPROVED** [review]
